Deriving from a custom class that implements Godot virtual methods fails to compile

With GodotCpp `master` branch.

Simple example:
```cpp
class TestNode : public godot::Node {
	GDCLASS(TestNode, godot::Node)
public:
	void _ready() override;
private:
	static void _bind_methods();
};

class DerivedNode : public TestNode {
	GDCLASS(DerivedNode, TestNode)
private:
	static void _bind_methods();
};
```

A setup like this one produces the following compiling errors:
```
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2672: 'call_with_ptr_args': no matching overloaded function found
TESTS\Godot4\Extension\addons\extension_test\cpp\src\test_node.h(9): note: see reference to function template instantiation 'void godot::Node::register_virtuals<T>(void)' being compiled
        with
        [
            T=extension_test::DerivedNode
        ]
TESTS\Godot4\Extension\addons\extension_test\cpp\src\derived_node.h(9): note: see reference to function template instantiation 'void extension_test::TestNode::register_virtuals<extension_test::DerivedNode>(void)' being compiled
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2782: 'void godot::call_with_ptr_args(T *,R (__cdecl T::* )(P...) const,const GDNativeTypePtr *,void *)': template parameter 'T' is ambiguous
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(207): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: could be 'extension_test::TestNode'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: or       'T'
        with
        [
            T=extension_test::DerivedNode
        ]
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2784: 'void godot::call_with_ptr_args(T *,R (__cdecl T::* )(P...) const,const GDNativeTypePtr *,void *)': could not deduce template argument for 'R (__cdecl T::* )(P...) const' from 'void (__cdecl extension_test::TestNode::* )(void)'
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(207): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2782: 'void godot::call_with_ptr_args(T *,R (__cdecl T::* )(P...),const GDNativeTypePtr *,void *)': template parameter 'T' is ambiguous
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(202): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: could be 'extension_test::TestNode'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: or       'T'
        with
        [
            T=extension_test::DerivedNode
        ]
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2784: 'void godot::call_with_ptr_args(T *,R (__cdecl T::* )(P...),const GDNativeTypePtr *,void *)': could not deduce template argument for 'R (__cdecl T::* )(P...)' from 'void (__cdecl extension_test::TestNode::* )(void)'
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(202): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2782: 'void godot::call_with_ptr_args(T *,void (__cdecl T::* )(P...) const,const GDNativeTypePtr *,void *)': template parameter 'T' is ambiguous
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(197): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: could be 'extension_test::TestNode'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: or       'T'
        with
        [
            T=extension_test::DerivedNode
        ]
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2784: 'void godot::call_with_ptr_args(T *,void (__cdecl T::* )(P...) const,const GDNativeTypePtr *,void *)': could not deduce template argument for 'void (__cdecl T::* )(P...) const' from 'void (__cdecl extension_test::TestNode::* )(void)'
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(197): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2782: 'void godot::call_with_ptr_args(T *,void (__cdecl T::* )(P...),const GDNativeTypePtr *,void *)': template parameter 'T' is ambiguous
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(192): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: could be 'extension_test::TestNode'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: or       'T'
        with
        [
            T=extension_test::DerivedNode
        ]
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2784: 'void godot::call_with_ptr_args(T *,void (__cdecl T::* )(P...),const GDNativeTypePtr *,void *)': could not deduce template argument for 'void (__cdecl T::* )(P...)' from 'void (__cdecl extension_test::TestNode::* )(void)'
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(192): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2672: 'call_with_ptr_args': no matching overloaded function found
TESTS\Godot4\Extension\addons\extension_test\cpp\src\test_node.h(9): note: see reference to function template instantiation 'void godot::Node::register_virtuals<T>(void)' being compiled
        with
        [
            T=extension_test::DerivedNode
        ]
TESTS\Godot4\Extension\addons\extension_test\cpp\src\derived_node.h(9): note: see reference to function template instantiation 'void extension_test::TestNode::register_virtuals<extension_test::DerivedNode>(void)' being compiled
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2782: 'void godot::call_with_ptr_args(T *,R (__cdecl T::* )(P...) const,const GDNativeTypePtr *,void *)': template parameter 'T' is ambiguous
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(207): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: could be 'extension_test::TestNode'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: or       'T'
        with
        [
            T=extension_test::DerivedNode
        ]
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2784: 'void godot::call_with_ptr_args(T *,R (__cdecl T::* )(P...) const,const GDNativeTypePtr *,void *)': could not deduce template argument for 'R (__cdecl T::* )(P...) const' from 'void (__cdecl extension_test::TestNode::* )(void)'
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(207): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2782: 'void godot::call_with_ptr_args(T *,R (__cdecl T::* )(P...),const GDNativeTypePtr *,void *)': template parameter 'T' is ambiguous
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(202): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: could be 'extension_test::TestNode'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: or       'T'
        with
        [
            T=extension_test::DerivedNode
        ]
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2784: 'void godot::call_with_ptr_args(T *,R (__cdecl T::* )(P...),const GDNativeTypePtr *,void *)': could not deduce template argument for 'R (__cdecl T::* )(P...)' from 'void (__cdecl extension_test::TestNode::* )(void)'
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(202): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2782: 'void godot::call_with_ptr_args(T *,void (__cdecl T::* )(P...) const,const GDNativeTypePtr *,void *)': template parameter 'T' is ambiguous
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(197): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: could be 'extension_test::TestNode'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: or       'T'
        with
        [
            T=extension_test::DerivedNode
        ]
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2784: 'void godot::call_with_ptr_args(T *,void (__cdecl T::* )(P...) const,const GDNativeTypePtr *,void *)': could not deduce template argument for 'void (__cdecl T::* )(P...) const' from 'void (__cdecl extension_test::TestNode::* )(void)'
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(197): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2782: 'void godot::call_with_ptr_args(T *,void (__cdecl T::* )(P...),const GDNativeTypePtr *,void *)': template parameter 'T' is ambiguous
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(192): note: see declaration of 'godot::call_with_ptr_args'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: could be 'extension_test::TestNode'
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): note: or       'T'
        with
        [
            T=extension_test::DerivedNode
        ]
Engine\godot_cpp_fork\gen\include\godot_cpp/classes/node.hpp(254): error C2784: 'void godot::call_with_ptr_args(T *,void (__cdecl T::* )(P...),const GDNativeTypePtr *,void *)': could not deduce template argument for 'void (__cdecl T::* )(P...)' from 'void (__cdecl extension_test::TestNode::* )(void)'
Engine\godot_cpp_fork\include\godot_cpp/core/binder_common.hpp(192): note: see declaration of 'godot::call_with_ptr_args'
scons: *** [src\derived_node.windows.debug.x86_64.obj] Error 2
scons: *** [src\register_types.windows.debug.x86_64.obj] Error 2
```

The only workaround I could find so far was to implement those virtual methods again in all derived classes, making them call the parent method. This is inconvenient and becomes a nightmare in cases like this https://github.com/Zylann/godot_voxel/blob/gdextension/editor/vox/vox_scene_importer.h

MRP:
[extension_test_issue854_virtual_methods.zip](https://github.com/godotengine/godot-cpp/files/9601389/extension_test_issue854_virtual_methods.zip)
Note: change the path to GodotCpp in `SConstruct`.

--------

It seems to happen because `GDCLASS` internally defines `register_virtuals()`, which calls `BaseClass::register_virtuals<CurrentClass>()`. And in there, it does this:
```cpp
		// Checks if `T` overrides `_ready`. If so, bind it.
		if constexpr (!std::is_same_v<decltype(&Node::_ready), decltype(&T::_ready)>) {
			BIND_VIRTUAL_METHOD(T, _ready);
		}
```
When `T` is the derived class, it does not define `_ready`. Only its base class does. But for some reason, that makes `BIND_VIRTUAL_METHOD` fail to compile.

I thought that's because the `if` checks the wrong condition. Indeed, when `T` derives from a custom class, `Node::_ready` is not the same function as `T::_ready`. But `T` does not actually define `_ready`, its custom base class does. So we should not try to bind the method again if it's the same as the base class.
So I tried this:
```cpp
		if constexpr (!std::is_same_v<decltype(&B::_ready), decltype(&T::_ready)>) {
			BIND_VIRTUAL_METHOD(T, _ready);
		}
```
Where `B` is the actual base class, so a more informed check can be done.
Unfortunately, the fact there is an `if` does not prevent the code inside from being compiled. The error remains exactly the same: it can't compile the inside of the `if`.

Here with the macro expanded:
```cpp
		if constexpr (!std::is_same_v<decltype(&B::_ready), decltype(&T::_ready)>) {
			// BIND_VIRTUAL_METHOD(T, _ready);
			auto ___call_ready = [](GDNativeObjectPtr p_instance, const GDNativeTypePtr *p_args, GDNativeTypePtr p_ret) -> void {
				call_with_ptr_args(reinterpret_cast<T *>(p_instance), &T::_ready, p_args, p_ret);
			};
			godot::ClassDB::bind_virtual_method(T::get_class_static(), "_ready", ___call_ready);
		}
```

So what I dont understand is, why all the methods I don't override are compiling just fine, while they should be in the same situation? 
With `if constexpr`, if the condition doesn't pass, the code inside isn't even compiled. But here it looks like it tries to compile it anyways.

Funny enough, if I replace the condition with 
```cpp
		if constexpr (false) {
```
It compiles fine. When I try `!std::is_same_v<decltype(&B::_ready), decltype(&T::_ready)>`, I tested and expected it to be `false` too for the derived class, but somehow discarding isn't working.
I even tried somthing similar on `https://godbolt.org/`  and it worked... 
![image](https://user-images.githubusercontent.com/1311555/191123545-66950f5f-d852-4f8b-b862-2e56135f25f6.png)

--- a/binding_generator.py
+++ b/binding_generator.py
@@ -1071,17 +1071,22 @@ def generate_engine_class_header(class_api, used_classes, fully_used_classes, us
             result.append(method_signature + ";")
 
     result.append("protected:")
-    result.append("\ttemplate <class T>")
+    # T is the custom class we want to register (from which the call initiates, going up the inheritance chain),
+    # B is its base class (can be a custom class too, that's why we pass it).
+    result.append("\ttemplate <class T, class B>")
     result.append("\tstatic void register_virtuals() {")
     if class_name != "Object":
-        result.append(f"\t\t{inherits}::register_virtuals<T>();")
+        result.append(f"\t\t{inherits}::register_virtuals<T, B>();")
         if "methods" in class_api:
             for method in class_api["methods"]:
                 if not method["is_virtual"]:
                     continue
                 method_name = escape_identifier(method["name"])
                 result.append(
-                    f"\t\tif constexpr (!std::is_same_v<decltype(&{class_name}::{method_name}),decltype(&T::{method_name})>) {{"
+                    # If the method is different from the base class, it means T overrides it, so it needs to be bound.
+                    # Note that with an `if constexpr`, the code inside the `if` will not even be compiled if the
+                    # condition returns false (in such cases it can't compile due to ambiguity).
+                    f"\t\tif constexpr (!std::is_same_v<decltype(&B::{method_name}),decltype(&T::{method_name})>) {{"
                 )
                 result.append(f"\t\t\tBIND_VIRTUAL_METHOD(T, {method_name});")
                 result.append("\t\t}")

--- a/include/godot_cpp/classes/wrapped.hpp
+++ b/include/godot_cpp/classes/wrapped.hpp
@@ -145,9 +145,9 @@ protected:                                                                      
 		return (::godot::String(::godot::Wrapped::*)()) & m_class::_to_string;                                                                           \
 	}                                                                                                                                                    \
                                                                                                                                                          \
-	template <class T>                                                                                                                                   \
+	template <class T, class B>                                                                                                                          \
 	static void register_virtuals() {                                                                                                                    \
-		m_inherits::register_virtuals<T>();                                                                                                              \
+		m_inherits::register_virtuals<T, B>();                                                                                                           \
 	}                                                                                                                                                    \
                                                                                                                                                          \
 public:                                                                                                                                                  \
@@ -159,7 +159,7 @@ public:                                                                         
 		m_inherits::initialize_class();                                                                                                                  \
 		if (m_class::_get_bind_methods() != m_inherits::_get_bind_methods()) {                                                                           \
 			_bind_methods();                                                                                                                             \
-			m_inherits::register_virtuals<m_class>();                                                                                                    \
+			m_inherits::register_virtuals<m_class, m_inherits>();                                                                                        \
 		}                                                                                                                                                \
 		initialized = true;                                                                                                                              \
 	}                                                                                                                                                    \

--- a/include/godot_cpp/core/class_db.hpp
+++ b/include/godot_cpp/core/class_db.hpp
@@ -89,10 +89,12 @@ public:
 		std::unordered_map<std::string, GDNativeExtensionClassCallVirtual> virtual_methods;
 		std::set<std::string> property_names;
 		std::set<std::string> constant_names;
+		// Pointer to the parent custom class, if any. Will be null if the parent class is a Godot class.
 		ClassInfo *parent_ptr = nullptr;
 	};
 
 private:
+	// This may only contain custom classes, not Godot classes
 	static std::unordered_map<std::string, ClassInfo> classes;
 
 	static MethodBind *bind_methodfi(uint32_t p_flags, MethodBind *p_bind, const MethodDefinition &method_name, const void **p_defs, int p_defcount);
@@ -158,10 +160,12 @@ void ClassDB::register_class() {
 	cl.name = T::get_class_static();
 	cl.parent_name = T::get_parent_class_static();
 	cl.level = current_level;
-	classes[cl.name] = cl;
-	if (classes.find(cl.parent_name) != classes.end()) {
-		cl.parent_ptr = &classes[cl.parent_name];
+	std::unordered_map<std::string, ClassInfo>::iterator parent_it = classes.find(cl.parent_name);
+	if (parent_it != classes.end()) {
+		// Assign parent if it is also a custom class
+		cl.parent_ptr = &parent_it->second;
 	}
+	classes[cl.name] = cl;
 
 	// Register this class with Godot
 	GDNativeExtensionClassCreationInfo class_info = {

--- a/src/core/class_db.cpp
+++ b/src/core/class_db.cpp
@@ -277,20 +277,29 @@ void ClassDB::bind_integer_constant(const char *p_class_name, const char *p_enum
 }
 
 GDNativeExtensionClassCallVirtual ClassDB::get_virtual_func(void *p_userdata, const char *p_name) {
+	// This is called by Godot the first time it calls a virtual function, and it caches the result, per object instance.
+	// Because of this, it can happen from different threads at once.
+	// It should be ok not using any mutex as long as we only READ data.
+
 	const char *class_name = (const char *)p_userdata;
 
 	std::unordered_map<std::string, ClassInfo>::iterator type_it = classes.find(class_name);
 	ERR_FAIL_COND_V_MSG(type_it == classes.end(), nullptr, "Class doesn't exist.");
 
-	ClassInfo &type = type_it->second;
+	const ClassInfo *type = &type_it->second;
 
-	std::unordered_map<std::string, GDNativeExtensionClassCallVirtual>::iterator method_it = type.virtual_methods.find(p_name);
+	// Find method in current class, or any of its parent classes (Godot classes not included)
+	while (type != nullptr) {
+		std::unordered_map<std::string, GDNativeExtensionClassCallVirtual>::const_iterator method_it = type->virtual_methods.find(p_name);
 
-	if (method_it == type.virtual_methods.end()) {
-		return nullptr;
+		if (method_it != type->virtual_methods.end()) {
+			return method_it->second;
+		}
+
+		type = type->parent_ptr;
 	}
 
-	return method_it->second;
+	return nullptr;
 }
 
 void ClassDB::bind_virtual_method(const char *p_class, const char *p_method, GDNativeExtensionClassCallVirtual p_call) {

